### PR TITLE
feat(operator): support MultiNamespace install mode

### LINKED ISSUE #644: Support MultiNamespace install mode
##### What can be improved

A single operator install should be able to watch multiple namespaces

##### Why

Less  resource usage

### PR DESCRIPTION
**Pull request by aslakknutsen**
#### Short description of what this resolves:

Allow for selecting Multi Namespace install mode in OLM.

#### Changes proposed in this pull request:

Move away from `Namespace`field on Manager and adjust the Cache directly.

Fixes #644

MultiNamespace tests are disabled for now due to https://github.com/operator-framework/operator-sdk/issues/4512

### CODE REVIEW

**Comment by alien-ike** [comment]
Skipping CI for Draft Pull Request.
If you want CI signal for your change, please convert it to an actual PR.
You can still manually trigger a test run with `/test all`

**Comment by alien-ike** [comment]
@aslakknutsen: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-tests-openshift | 309de1dd99dc7501efb30b2dc93a15a14186f892 | [link](http://status-prow.apps.workspace.maistra.test/view/gs/ike-prow-logs/pr-logs/pull/maistra_istio-workspace/680/e2e-tests-openshift/1359990504757923840) | `/test openshift`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by alien-ike** [comment]
@aslakknutsen: The following test **failed**, say `/retest` to rerun all failed tests:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-tests-openshift | e5d6ad1149f6193eb0752df6114e4f20957405cb | [link](http://status-prow.apps.workspace.maistra.test/view/gs/ike-prow-logs/pr-logs/pull/maistra_istio-workspace/680/e2e-tests-openshift/1360171950583844864) | `/test openshift`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: e2e/install_mode_test.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+		WatchListExpression := func() string {
+			return strings.Join(namespaces, ",")
+		}
+		VerifyWatchList := func(validateNamespaces ...string) {
+			for _, watchNs := range validateNamespaces {
+				ikeCreate := RunIke(shell.GetProjectDir(), "create",
+					"--deployment", watchNs+"-v1",
+					"-n", watchNs,
+					"--route", "header:x-test-suite=smoke",
+					"--image", "x:x:x", // never used
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:56]: Maybe worth explaining a bit more in the comment why we have this "dummy" deployment at all :) We might forget at one point.

File: e2e/install_mode_test.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+			VerifyWatchList(namespaces...)
+		})
+		It("AllNamespace", func() {
+			namespaces = append(namespaces, generateNamespaceName(), generateNamespaceName())
+			CreateNamespace()
+
+			bundle := shell.ExecuteInDir(shell.GetProjectDir(), "make", "bundle-run-all")
+			<-bundle.Done()
+			Expect(bundle.Status().Exit).To(Equal((0)))
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:124]: Apart of `Equal` having one too many `()`, we can also say:

```suggestion
			Expect(bundle.Status().Exit).To(BeZero())
```

File: e2e/install_mode_test.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+		PIt("MultiNamespace", func() { // require Operator SDK update
+			namespaces = append(namespaces, generateNamespaceName(), generateNamespaceName())
+			CreateNamespace()
+
+			defer test.TemporaryEnvVars("OPERATOR_WATCH_NAMESPACE", WatchListExpression())()
+
+			bundle := shell.ExecuteInDir(shell.GetProjectDir(), "make", "bundle-run-multi")
+			<-bundle.Done()
+			Expect(bundle.Status().Exit).To(Equal((0)))
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:112]: Apart of `Equal` having one too many `()`, we can also say:

```suggestion
			Expect(bundle.Status().Exit).To(BeZero())
```

File: e2e/install_mode_test.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+		It("SingleNamespace", func() {
+			namespaces = append(namespaces, generateNamespaceName())
+			CreateNamespace()
+
+			defer test.TemporaryEnvVars("OPERATOR_WATCH_NAMESPACE", WatchListExpression())()
+
+			bundle := shell.ExecuteInDir(shell.GetProjectDir(), "make", "bundle-run-single")
+			<-bundle.Done()
+			Expect(bundle.Status().Exit).To(Equal((0)))
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:97]: Apart of `Equal` having one too many `()`, we can also say:

```suggestion
			Expect(bundle.Status().Exit).To(BeZero())
```

File: e2e/install_mode_test.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+				ikeCreate.Stop()
+				<-shell.ExecuteInDir(".", "kubectl", "delete", "session", watchNs, "-n", watchNs).Done()
+			}
+
+		}
+		It("OwnNamespace", func() {
+			bundle := shell.ExecuteInDir(shell.GetProjectDir(), "make", "bundle-run")
+			<-bundle.Done()
+			Expect(bundle.Status().Exit).To(Equal((0)))
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:82]: Apart of `Equal` having one too many `()`, we can also say:

```suggestion
			Expect(bundle.Status().Exit).To(BeZero())
```

File: e2e/install_mode_test.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+						return strings.Contains(log, contain)
+					}
+				}(watchNs), 1*time.Minute, 5*time.Second).Should(BeTrue())
+
+				ikeCreate.Stop()
+				<-shell.ExecuteInDir(".", "kubectl", "delete", "session", watchNs, "-n", watchNs).Done()
+			}
+
+		}
+		It("OwnNamespace", func() {
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:79]:
```suggestion
		It("should install to its own namespace", func() {
```

File: e2e/install_mode_test.go [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+			bundle := shell.ExecuteInDir(shell.GetProjectDir(), "make", "bundle-run")
+			<-bundle.Done()
+			Expect(bundle.Status().Exit).To(Equal((0)))
+
+			Eventually(AllDeploymentsAndPodsReady(operatorNamepace), 10*time.Minute, 5*time.Second).Should(BeTrue())
+
+			VerifyWatchList(operatorNamepace)
+		})
+
+		It("SingleNamespace", func() {
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:89]:
```suggestion
		It("should install to the single namespace", func() {
```

File: e2e/install_mode_test.go [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+			bundle := shell.ExecuteInDir(shell.GetProjectDir(), "make", "bundle-run-single")
+			<-bundle.Done()
+			Expect(bundle.Status().Exit).To(Equal((0)))
+
+			Eventually(AllDeploymentsAndPodsReady(operatorNamepace), 10*time.Minute, 5*time.Second).Should(BeTrue())
+
+			VerifyWatchList(namespaces...)
+		})
+
+		PIt("MultiNamespace", func() { // require Operator SDK update
```
Comment (bartoszmajsak) [inline e2e/install_mode_test.go:104]:
```suggestion
		PIt("should install to multiple namespaces", func() { // require Operator SDK update, see: https://github.com/operator-framework/operator-sdk/issues/4512
```

**Review by bartoszmajsak — CHANGES_REQUESTED** [review]
\o/

Few minor suggestions with regards to test style / naming, but other than that looks good.

There's one question though - about the CSV setting saying we support `MultiNamespace`. See my comments below :)

File: config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml [hunk 9]
```diff
@@ -29,7 +29,7 @@ spec:
     type: OwnNamespace
   - supported: true
     type: SingleNamespace
-  - supported: false
+  - supported: true
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml:32]: It's been tested manually and via the fixed operator-sdk v.

We could of course split out the test suite from this PR and only commit the MultiNamespace part when the Operator-SDK 'test infra' is updated... but....

File: e2e/install_mode_test.go (hunk 1 above)
Comment (aslakknutsen) [reply e2e/install_mode_test.go:74]: https://github.com/maistra/istio-workspace/pull/680/commits/8945c58858762b4f17a366eb4fc938ce18b869fa

File: config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply config/manifests/bases/istio-workspace-operator.clusterserviceversion.yaml:32]: We are confident that works as intended then sure - we can keep that enabled

**Review by bartoszmajsak — APPROVED** [review]

**Pull request merged**